h2o-algos/src/main/java/hex/glm/ComputationState.java

```diff
@@ -570,6 +568,21 @@ public GramXY(Gram gram, double[] xy, double [] grads, double[] beta, int[] acti
       this.likelihood = likelihood;
     }
 
+    public final double [] getCODGradients(double l2pen){
+      int lastGradInd = xy.length-1;
+      if(grads == null){
+        grads = new double[xy.length];
+        for(int i = 0; i < lastGradInd; ++i)
+          grads[i] = xy[i] - l2pen * beta[i];
+        grads[lastGradInd] = xy[lastGradInd]; // intercept does not have l1/l2 penality
+      }
+      if(newCols != null) {
+        for (int i : newCols)
+          grads[i] = xy[i] - (i==lastGradInd?0:l2pen * beta[i]);
+      }
+      return grads;
+    }
```

Comment:
This is different from getCODGradients(). Here, I am trying to calculate the gradient as the partial derivatives of the negative log likelihood + l2pen and I need this to update coefficients for COD. This is equaition 4.24 of ESLII-print4 without the l2/l1 penalty